### DISCUSSION TITLE
효율적인 모델 파인 튜닝 전략

### DISCUSSION THREAD

**Discussion by 4N3MONE** [Pseudo-Lab/Hugging-Face-Hub-Garden#34] [Seminar Materials]
# 효율적인 모델 파인 튜닝 전략

### 파인 튜닝의 기본 이해

- 파인 튜닝이란? 개념 및 중요성

![image](https://github.com/user-attachments/assets/fb057385-ecb3-4f98-9269-7bf751a20cfc)

**파인튜닝(fine-tuning, 미세 조정) 이란** 

파인튜닝은 언어 모델을 특정 작업에 대해서 파라미터 재조정을 통해 훈련시키는 것을 의미합니다.

파인튜닝을 하기 위해서는 **‘사전학습’**된 언어 모델이 필수적으로 필요합니다. 비유하자면 사전학습은 말과 글을 사용하는 법을 가르쳐주는 것이고, 파인튜닝은 기사문을 작성하는 법, 통역을 하는 법 등 구체적인 사용 사례를 가르쳐주는 것입니다. 

초기 언어 모델(ELMO, BERT) 패러다임에서는 파인튜닝을 실시할 때에 마지막 레이어에 태스크 종류에 따른 레이어를 추가하여 모델을 파인튜닝 하는 것이 일반적이었으나, gpt류의 디코더 모델들이 주류가 되면서 **Instruction Tuning** 기법이 일반화 되었고, 본 포스트에서도 그것을 기준으로 다루도록 하겠습니다. 

**파인튜닝을 하기 위해 ‘꼭’ 필요한 것들.**

1. **사전 학습된 언어 모델**
    
    앞서 말했듯이 파인 튜닝은 언어 모델에게 언어, 글에 대한 지식을 학습시키는 것이 아니라 사용하는 방법을 학습시키는 것입니다. 따라서 파인튜닝을 할 때에는 반드시 사전학습된 언어 모델이 필요합니다. 이러한 모델들은 보통 ‘base’ 등의 suffix 를 붙여서 공개되는 경우가 많습니다. 
    
    *Q : 이미 파인튜닝된 모델을 한번 더 파인튜닝하면 안되나요?*
    
    A : *가능합니다.* 언어 모델들의 파라미터 크기가 점점 더 커짐에 따라 개인이 접근할 수 있는 계산 자원으로 언어 모델을 파인튜닝하는 일은 점점 어려워졌습니다. 따라서 언어 모델을 공개하는 측에서도 굳이 학습해서 파인튜닝해야하는 사전 학습 모델을 공개하는 것이 아닌, 한 번 파인튜닝을 한 모델을 공개하는 것이 대세가 되었습니다. 이런 경우에는 모델 이름에 ‘Instruction, INST’ 등의 suffix가 붙거나 `tokenizer_config`에 `chat_template` 가 포함되어 있습니다. 저 또한 파인튜닝을 하고싶은 경우 이러한 Instruction 모델에 파인 튜닝을 하는 것을 권장드립니다. 한글만 할 줄 아는 아이에게 논문 작성법을 가르치는 것보다, 고등학교 까지의 국어 수업을 마친 아이에게 가르치는 것이 쉬운 것과 같은 이치입니다. 
    
2. **학습시킬 데이터셋**
    
    파인튜닝으로 달성하고자 하는 목적에 걸맞는 데이터셋을 선택(제작)하는 것은 언어 모델 파인튜닝에 있어서 가장 중요한 작업 중 하나입니다. 파인튜닝된 언어 모델의 성능은 데이터셋의 **크기**와 **품질**에 크게 구애받습니다. 충분한 품질과 크기를 갖추지 못한 데이터셋으로 언어 모델을 파인튜닝 할 경우, 파인튜닝하기 전보다 성능이 나쁜 결과물이 나오게 되는 경험을 하게 됩니다.
    
3. **학습하기 위한 자원(GPU)**
    
    언어 모델을 학습하려면 GPU 자원이 필수적입니다. 밑에서 서술할 파라미터-효율적 파인 튜닝 방법론 (PEFT)의 등장에 따라, 대규모 GPU 클러스터 없이도 언어 모델 파인튜닝이 가능하게 되었습니다. 하지만, 가용 자원에서 최대한 좋은 성능을 낼 수 있는 방법을 선택하는 것이 좋으므로 적합한 학습 방법을 선택하는 것도 중요한 일입니다. 
    
    [언어 모델과 GPU 자원을 통해 학습 가능 여부를 가늠할 수 있는 HuggingFace Space](https://huggingface.co/spaces/Vokturz/can-it-run-llm)
    
    [종량제 GPU 대여 서비스 Runpod(이외에도 많음)](https://www.runpod.io/)
    

**파인튜닝에서 있으면 좋은 것들.**

1. 평가 메트릭
    
    시간, 자원을 들여 언어 모델을 학습시켰는데 학습하기 이전보다 성능이 안좋은 것은 너무나 슬픈 일입니다. 이러한 일을 방지하기 위해 내가 원하는 작업을 수행할 수 있는지 측정하는 평가 지표를 설정할 수 있으면 좋습니다. 보통 다음과 같은 방법들이 권장됩니다.
    
    - 기존 벤치마크 데이터셋을 사용하기
    - 다른 모델과의 블라인드 테스트(Model Arena)
    - 훈련 데이터셋과 분리된 Validation Set으로 평가
    - 실제 유즈케이스와 유사한 테스트 셋 만들어서 평가

1. 학습 파이프라인
    
    언어 모델 학습은 보통 수 시간~수 일까지 소요되는 작업이고, 작은 파라미터의 차이로 결과물의 차이가 크게 달라질 수 있는 섬세한 작업이기도 합니다. 따라서 학습 과정을 **모니터링**하면서 추이를 관찰하고, 학습 코드의 각 부분을 **모듈화**하여 수정이 용이하게 하는 것이 좋습니다. 모니터링에는 다음과 같은 Tool들이 많이 사용됩니다. 
    
    [mlflow](https://github.com/mlflow/mlflow)
    
    [wandb](https://github.com/wandb/wandb)
    

이러한 학습 파이프라인 툴들은 하루종일 터미널 창을 들여다보지 않더라도 학습 진행상황 등을 기록하여 이후에 확인할 수 있도록 해주고, 재현성을 가질 수 있도록 큰 도움을 줍니다. 

![image 1](https://github.com/user-attachments/assets/568d2b90-1932-4222-ad2d-4c7b475e2efd)

### 여기서 잠깐!  Fine-Tuning과 Instruction Tuning, Chat Template에 관하여

언어 모델을 파인튜닝하다보면 꼭 마주치게 되는 것이 있습니다. 바로 Chat Template 라는 것인데, 다음 챕터로 넘어가기 전에 이것에 대해서 알아보도록 하겠습니다. 

디코더 기반 언어 모델은 사용자, 어시스턴트, 시스템 등의 역할을 명확히 구분하고 멀티턴 대화에서 문맥을 정확히 유지하기 위해 Chat Template를 사용합니다. 이 템플릿은 대화의 각 요소를 **고정된 형식으로 구조화**하며, 일반적으로 `<|user|>`, `<|assistant|>`와 같은 special token을 포함하여 **역할별 발화**를 명시적으로 표현합니다. 

```python
from transformers import AutoTokenizer

# 입력 메시지
messages = [
    {"role": "system", "content": "You are a helpful assistant."},
    {"role": "user", "content": "서울의 수도는 어디야?"},
]

# Chat Template 적용
prompt = tokenizer.apply_chat_template(messages, tokenize=False)
print(prompt)

"""
<|system|>
You are a helpful assistant.
<|user|>
서울의 수도는 어디야?
<|assistant|>
"""
```

실제로 언어 모델이 학습 시, 추론 시 사용하는 텍스트 시퀀스는 출력된 내용과 같으며,  채팅 서비스와 같은 실제 어플리케이션에서는 역할과 턴 구분등에 사용되는 스페셜 토큰을 기반으로 UI에 출력하게 됩니다.

### 효율적인 파인 튜닝 기법 소개

**Parameter-Efficient Fine-Tuning (PEFT)**

![image 2](https://github.com/user-attachments/assets/b5ab4261-a55e-477e-8c68-c4045aff9a58)

언어 모델의 모든 파라미터에 
대해서 학습하는 Full Fine-Tuning과 달리, Parameter-Efficient Fine-Tuning(PEFT)은 기존 모델의 대부분 파라미터를 고정하고 소수의 추가 파라미터만 학습함으로써 적은 자원으로도 효과적인 미세조정을 가능하게 하는 방법입니다. 

- LoRA(Low-Rank Adaptation)
    
    Low-Rank Adaptation(Lora)은 모델의 원래 파라미터를 고정하고, 특정 가중치 행렬에 대해 저랭크 행렬을 삽입해 학습 가능한 소규모 파라미터만 추가하는 방식으로 미세조정을 수행합니다. 
    
    ```python
    from peft import LoraConfig, TaskType
    
    lora_config = LoraConfig(
        r=8,  # LoRA 랭크 (low-rank 행렬의 차원). 작을수록 학습 파라미터 수가 줄어들며, 일반적으로 4~64 범위 사용.
        lora_alpha=32,  # LoRA scaling factor. LoRA의 출력을 얼마나 크게 반영할지 결정하는 계수.
        target_modules=["q_proj", "v_proj"],  # LoRA를 적용할 모델 내부 모듈 이름. 예: attention의 query, value projection 등.
        lora_dropout=0.05,  # 학습 중 LoRA 모듈에 적용할 dropout 확률. 과적합 방지를 위해 사용.
        bias="none",  # bias 파라미터를 학습할지 여부 ("none", "all", "lora_only" 중 선택).
        task_type=TaskType.CAUSAL_LM  # 해당 LoRA 설정이 적용될 태스크 타입. 예: CAUSAL_LM, SEQ_CLS, TOKEN_CLS 등.
    )
    
    ```
    
- QLoRA
    
    Quantized Low-Rank Adaptation(QLoRA)은 모델의 전체 파라미터를 4비트 등 저정밀도로 양자화하여 GPU 메모리 사용량을 줄인 상태에서, LoRA와 같은 저랭크 어댑터를 덧붙여 소규모 파라미터만 학습하는 방식으로 미세조정을 수행합니다.
    

```python
from transformers import AutoModelForCausalLM, AutoTokenizer, BitsAndBytesConfig
from peft import LoraConfig, get_peft_model, TaskType

# 1. 4bit 양자화 설정 (bitsandbytes)
bnb_config = BitsAndBytesConfig(
    load_in_4bit=True,            # 4비트 양자화 사용
    bnb_4bit_compute_dtype="float16",  # 내부 연산 정밀도 (보통 float16이나 bfloat16)
    bnb_4bit_quant_type="nf4",         # quantization 방식 (nf4가 QLoRA 논문에서 추천됨)
)

# 2. 모델과 토크나이저 로드 (4bit 양자화된 모델로)
model_name = "meta-llama/Llama-2-7b-hf"  # 원하는 모델 이름
model = AutoModelForCausalLM.from_pretrained(
    model_name,
    quantization_config=bnb_config, # 양자화해서 모델 로드
    device_map="auto"
)
tokenizer = AutoTokenizer.from_pretrained(model_name, use_fast=True)

# 3. QLoRA 설정
lora_config = LoraConfig(
    r=8,  # 저랭크 차원
    lora_alpha=16,  # LoRA 출력 스케일 계수
    target_modules=["q_proj", "v_proj"],  # LoRA 적용 대상 모듈
    lora_dropout=0.05,  # dropout 적용 (학습 시)
)

# 4. QLoRA 모델로 래핑
model = get_peft_model(model, lora_config)
```

### (효율적이진 않지만) 다양한 파인 튜닝 방법론 소개

- DPO(Direct Preference Optimization)

![image 3](https://github.com/user-attachments/assets/9eb42af6-4733-4d71-bda8-8c93b315c450)

(ChatGPT가 존재할 수 있도록 했던) PPO, RLHF 는 강화 학습을 활용해 언어 모델의 성능을 획기적으로 끌어 올렸지만 reward model의 학습이 필수적으로 필요하고, 학습이 stable 하지 않는 등 진입 장벽이 높은 방법입니다. DPO는 기존 지도학습의 형식을 따르면서도 RLHF의 장점을 취할 수 있도록 `chosen`, `reject`로 레이블링 된 데이터셋을 활용하여 선호되는 응답을 생성할 확률을 높이고, 비선호되는 응답을 생성할 확률을 낮추는 방식으로 학습을 진행합니다. 

```python
# train_dpo.py
from datasets import load_dataset
from trl import DPOConfig, DPOTrainer
from transformers import AutoModelForCausalLM, AutoTokenizer

model = AutoModelForCausalLM.from_pretrained("Qwen/Qwen2-0.5B-Instruct")
tokenizer = AutoTokenizer.from_pretrained("Qwen/Qwen2-0.5B-Instruct")
train_dataset = load_dataset("trl-lib/ultrafeedback_binarized", split="train")

training_args = DPOConfig(output_dir="Qwen2-0.5B-DPO", logging_steps=10)
trainer = DPOTrainer(model=model, args=training_args, processing_class=tokenizer, train_dataset=train_dataset)
trainer.train()
```

경험상 DPO는 SFT와 같은 ‘모델이 새로운 지식을 습득하는’ 효과는 얻어내지 못하지만, ‘모델의 행동을 내 의도대로 교정하는’ (code-mixing, sentence style) 데에는 좋은 성능을 보였습니다.

[코드 예시 전문(허깅페이스 블로그 포스트)](https://huggingface.co/blog/dpo-trl)

- GRPO(Group Relative Policy Optimization)

PPO의 Critic Model 없이, 같은 질문에 대해 모델이 생성한 여러 답변들의상대적 성능을 기준으로 학습하는 방법입니다. 이를 통해 자원 효율을 높이면서도 PPO와 유사한 학습 안정성과 성능 향상을 도모할 수 있으며, Clip 및 KL 제약은 그대로 유지하여 극단적 학습을 방지합니다.

```python
# train_grpo.py
from datasets import load_dataset
from trl import GRPOConfig, GRPOTrainer

dataset = load_dataset("trl-lib/tldr", split="train")

# Define the reward function, which rewards completions that are close to 20 characters
def reward_len(completions, **kwargs):
    return [-abs(20 - len(completion)) for completion in completions]

training_args = GRPOConfig(output_dir="Qwen2-0.5B-GRPO", logging_steps=10)
trainer = GRPOTrainer(
    model="Qwen/Qwen2-0.5B-Instruct",
    reward_funcs=reward_len,
    args=training_args,
    train_dataset=dataset,
)
trainer.train()
```

GRPO는 위의 예시 코드와 같이 **정량적으로 측정 가능한 reward function**이 필요하며, 언어 모델의 수학문제, 코딩 수행 능력 등 추론을 필요로 하는 학습에 많이 사용됩니다. 

[HuggingFace Documentaion](https://huggingface.co/docs/trl/v0.17.0/en/grpo_trainer#grpo-trainer)

### 효율적으로 파인 튜닝을 하기 위한 도구

![image 4](https://github.com/user-attachments/assets/e1028dd0-ba49-44a2-a1e9-3e113c5ed9f7)

triton 커널 최적화로 연산 시간을 매우 빠르게 단축시키고, GPU 자원을 절약합니다.

- [unsloth](https://github.com/unslothai/unsloth)

![image 5](https://github.com/user-attachments/assets/0862e3d2-b6bd-4289-83b2-a36e653b0904)

PEFT 한정, Single-GPU 한정이지만 큰 효율 상승을 보여주며, 예시 또한 다양하게 제공합니다.

다만 HuggingFace Trainer 와 다른 독자적 라이브러리를 사용해야 하며, dependency issue를 겪기 쉽습니다.

- [Accelerate](https://huggingface.co/docs/accelerate/index), [Deepspeed](https://github.com/deepspeedai/DeepSpeed)

![image 6](https://github.com/user-attachments/assets/95512b6d-719e-4e6a-8b34-bcaa7237b29c)

Multi-GPU 사용, CPU-Offload 등의 테크닉을 사용해야 할 때 사용하게 되는 라이브러리들입니다. 

(Accelerate에서 deepspeed의 대부분의 기능을 포함하고 있음)

**Comment by harheem** [comment]
가이드북을 위한 연관 자료:
https://huggingface.co/learn/llm-course/en/chapter11/1?fw=pt
